diamond_square.js

```diff
@@ -1,142 +1,384 @@
+        for (let i = 0; true; i++) {
+          if (i !== 0) {
+            curFactorial *= i
+            curExp *= this.expectedPoints
+          }
+          numPointsRandomNumber -= curExp / curFactorial
+          if (numPointsRandomNumber <= 0) {
+            numPoints = i
+            break
+          }
+        }
+        for (let i = 0; i < numPoints; i++) {
+          thisPoints.push([
+            centerX + centeredRng.intBetween(-this.batchSize / 2, this.batchSize / 2),
+            centerY + centeredRng.intBetween(-this.batchSize / 2, this.batchSize / 2)
+          ])
+        }
+        this.pointCache[centerX][centerY] = thisPoints
+      }
+      for (const point of this.pointCache[centerX][centerY]) {
+        points.push(point)
+      }
     }
+    // Now, get closest point
+    let closestPointIdx = -1
+    let minSqDist = Infinity
+    for (const pointXY of points) {
+      const dx = pointXY[0] - x
+      const dy = pointXY[1] - y
+      const sqDist = dx * dx + dy * dy
+      if (sqDist < minSqDist) {
+        minSqDist = sqDist
+        closestPointIdx = pointXY[0] + pointXY[1] // TODO: Better hash
+      }
+    }
+    // Now, get maximum distance
+    let maxDist = 0
+    for (let i = 0; i < points.length; i++) {
+      const point1X = points[i][0]
+      const point1Y = points[i][1]
+      for (let j = 0; j < i; j++) {
+        const point2X = points[j][0]
+        const point2Y = points[j][1]
+        const dx = point1X - point2X
+        const dy = point1Y - point2Y
+        const sqDist = dx * dx + dy * dy
+        if (maxDist < sqDist) {
+          maxDist = sqDist
+        }
+      }
+    }
+    return [closestPointIdx, minSqDist / Math.sqrt(maxDist)]
   }
 }
 
-function generation ({ version = '1.8', seed, worldHeight = 80, waterline = 20 } = {}) {
-  // Selected empirically
-  const size = 10000000
+function duplicateArr (arr, times) {
+  return Array(times).fill([...arr]).reduce((a, b) => a.concat(b))
+}
+
+function generation ({ version = '1.8', seed, worldHeight = 80, waterline = 32, size = 10000000, roughness = null } = {}) {
   const Chunk = require('prismarine-chunk')(version)
-  const mcData = require('minecraft-data')(version)
-  const majorVersion = mcData.version.majorVersion
-  const space = new DiamondSquare(size, size / 500, seed)
+  const registry = require('prismarine-registry')(version)
+
+  // Selected empirically
+  if (roughness === null) roughness = size / 500
+  const seedRand = rand.create(seed)
+  const maxInt = 2 ^ 53 - 1
+  const surfaceNoise = new Perlin(seedRand(0, maxInt))
+  const soilNoise = new Perlin(seedRand(0, maxInt))
+  const soilNoise2 = new Perlin(seedRand(0, maxInt))
+  const bedrockNoise = new Perlin(seedRand(0, maxInt))
+  const biomeNoise = new Worley(0.00005, seedRand(0, maxInt))
+
+  const biomes = [
+    ...duplicateArr(['plains'], 15),
+    ...duplicateArr(['forest'], 20),
+    ...duplicateArr(['desert'], 10)
+  ]
 
   function generateSimpleChunk (chunkX, chunkZ) {
```

Comment:
This overall looks a bit messy, all the copies, temporary objects and lookups will make generation substantially slower. Can you do a benchmark before vs after?

Reply:
Chunk generation when moving with an elytra is not visibly slower than the previous world gen, and uses less memory.

Reply:
perf tips based after the benchmark:
* don't define objects (use [] or {}) inside hot loops that need to run many times (this creates lots of garbage), instead put them at the top
* don't use long object.property.access as 3 different lookups need to happen opposed to one, avoid null checking (like x?.y) when possible
* avoid hash maps / don't check for features inside for loops
* use integers instead of strings when comparing things, which are also garbage that has to be copied around (js strings are by copy over reference)